File: app/src/main/java/com/vistony/salesforce/Dao/Retrofit/ReasonDispatchRepository.java
package com.vistony.salesforce.Dao.Retrofit;

import static com.vistony.salesforce.Controller.Utilitario.Utilitario.getDateTime;

import android.content.Context;

import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import com.vistony.salesforce.Controller.Retrofit.Api;
import com.vistony.salesforce.Controller.Retrofit.Config;
import com.vistony.salesforce.Dao.SQLite.BancoSQLite;
import com.vistony.salesforce.Dao.SQLite.ParametrosSQLite;
import com.vistony.salesforce.Dao.SQLite.ReasonDispatchSQLite;
import com.vistony.salesforce.Entity.Retrofit.Modelo.ReasonDispatchEntity;
import com.vistony.salesforce.Entity.Retrofit.Modelo.TypeDispatchEntity;
import com.vistony.salesforce.Entity.Retrofit.Respuesta.ReasonDispatchEntityResponse;
import com.vistony.salesforce.Entity.Retrofit.Respuesta.TypeDispatchEntityResponse;

import java.util.List;
import java.util.Objects;
import java.util.concurrent.Executor;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class ReasonDispatchRepository  extends ViewModel {
    private MutableLiveData<String> status= new MutableLiveData<>();
    ReasonDispatchSQLite reasonDispatchSQLite;
    ParametrosSQLite parametrosSQLite;
    public MutableLiveData<String> geReasonDispatch (String Imei, Context context, Executor executor){
        Config.getClient().create(Api.class).getReasonDispatch(Imei).enqueue(new Callback<ReasonDispatchEntityResponse>() {
            @Override
            public void onResponse(Call<ReasonDispatchEntityResponse> call, Response<ReasonDispatchEntityResponse> response) {

                ReasonDispatchEntityResponse reasonDispatchEntityResponse=response.body();
                if(response.isSuccessful() && Objects.requireNonNull(reasonDispatchEntityResponse).getReasonDispatchEntities().size()>0){
                    executor.execute(() -> {
                    reasonDispatchSQLite = new ReasonDispatchSQLite(context);
                    parametrosSQLite = new ParametrosSQLite(context);
                    reasonDispatchSQLite.DeleteTableReasonDispatch();
                    reasonDispatchSQLite.AddReasonDispatch (reasonDispatchEntityResponse.getReasonDispatchEntities());
                    Integer countReasonDispatch=getCountReasonDispatch(context);
                    parametrosSQLite.ActualizaCantidadRegistros("24", "MOTIVO DESPACHO", ""+countReasonDispatch, getDateTime());
                    status.setValue("1");
                    });
                }else
                {
                    status.setValue("0");
                }
            }

            @Override
            public void onFailure(Call<ReasonDispatchEntityResponse> call, Throwable t) {
                status.setValue("0");
            }
        });
        return status;
    }

    private Integer getCountReasonDispatch(Context context){
        return reasonDispatchSQLite.getCountReasonDispatch() ;
    }
}

File: app/src/main/java/com/vistony/salesforce/kotlin/Model/ReasonDispatchRepository.kt
package com.vistony.salesforce.kotlin.Model

import android.content.Context
import android.util.Log
import androidx.lifecycle.MutableLiveData
import com.vistony.salesforce.kotlin.Utilities.api.RetrofitApi
import com.vistony.salesforce.kotlin.Utilities.api.RetrofitConfig
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors

class ReasonDispatchRepository {
    private val _status = MutableLiveData<String>()
    val status: MutableLiveData<String> = _status

    private val _result_get = MutableStateFlow(ResponseReasonDispatch())
    val result_get: StateFlow<ResponseReasonDispatch> get() = _result_get

    suspend fun addReasonDispatch(Imei:String,context: Context)
    {
        Log.e(
            "REOS",
            "ReasonDispatchRepository-getReasonDispatch-Imei"+Imei
        )

        try {
            val retrofitConfig: RetrofitConfig? = RetrofitConfig()
            val service = retrofitConfig?.getClientLog()?.create(
                RetrofitApi
                ::class.java
            )

            Log.e(
                "REOS",
                "ReasonDispatchRepository-getReasonDispatch-service" + service.toString()
            )
            service?.getReasonDispatch(
                Imei
            )?.enqueue(object : Callback<ReasonDispatchResponse?> {

                override fun onResponse(
                    call: Call<ReasonDispatchResponse?>,
                    response: Response<ReasonDispatchResponse?>
                ) {
                    Log.e(
                        "REOS",
                        "ReasonDispatchRepository-getReasonDispatch-call$call"
                    )

                    Log.e(
                        "REOS",
                        "ReasonDispatchRepository-getReasonDispatch-response$response"
                    )

                    Log.e(
                        "REOS",
                        "ReasonDispatchRepository-getReasonDispatch-response.isSuccessful"
                    )
                    val reasonDispatchResponse = response.body()
                    Log.e(
                        "REOS",
                        "ReasonDispatchRepository?.getReasonDispatch()?.size" + reasonDispatchResponse?.getReasonDispatch()?.size
                    )

                    if (response.isSuccessful&&reasonDispatchResponse?.getReasonDispatch()?.size!!>0) {
                        val executor: ExecutorService = Executors.newFixedThreadPool(1)
                        for (i in 1..1) {
                            executor.execute {
                                println("Tarea $i en ejecución en ${Thread.currentThread().name}")
                                //Thread.sleep(1000)
                                val database by lazy { AppDatabase.getInstance(context.applicationContext) }
                                database?.reasonDispatchDao
                                    ?.deleteReasonDispatch()
                                database?.reasonDispatchDao
                                    ?.inserReasonDispatch (
                                        reasonDispatchResponse?.getReasonDispatch()
                                    )

                                println("Tarea $i completada")
                            }

                        }

                        executor.shutdown()
                        _status.setValue("1")
                        Log.e(
                            "REOS",
                            "ReasonDispatchRepository-getReasonDispatch-response.isSuccessful.status" + _status.getValue()
                        )
                    } else {
                        Log.e(
                            "REOS",
                            "ReasonDispatchRepository-getReasonDispatch-response.isSuccessful"
                        )
                        _status.setValue("0")
                        Log.e(
                            "REOS",
                            "ReasonDispatchRepository-getReasonDispatch-response.isSuccessful.not.status" + _status.getValue()
                        )
                    }
                    Log.e(
                        "REOS",
                        "ReasonDispatchRepository-getReasonDispatch-response.noentroenif.status" + _status.getValue()
                    )

                }

                override fun onFailure(call: Call<ReasonDispatchResponse?>, t: Throwable) {
                    _status.setValue("0")

                    Log.e(
                        "REOS",
                        "ReasonDispatchRepository-getReasonDispatch-error$t"
                    )
                }
            })


        } catch (e: Exception) {
            Log.e(
                "REOS",
                "ReasonDispatchRepository-getReasonDispatch-error: " + e.toString()
            )
        }
    }

    fun getReasonDispatch(
        context: Context,
    )
    {
        try {
            val executor: ExecutorService = Executors.newFixedThreadPool(1)
            for (i in 1..1) {
                executor.execute {
                    println("Tarea $i en ejecución en ${Thread.currentThread().name}")
                    //Thread.sleep(1000)
                    val database by lazy { AppDatabase.getInstance(context.applicationContext) }

                    val data=database?.reasonDispatchDao
                        ?.getReasonDispatch()
                    _result_get.value=ResponseReasonDispatch(Status = "Y",data = data)
                    Log.e(
                        "REOS",
                        "TypeDispatchRepository-getTypeDispatch-data: " + data
                    )
                    Log.e(
                        "REOS",
                        "TypeDispatchRepository-getTypeDispatch-_result_get.value.data: " + _result_get.value.data
                    )
                    println("Tarea $i completada")
                }
            }
            executor.shutdown()
        }
        catch (e: Exception) {
            Log.e(
                "REOS",
                "TypeDispatchRepository-getTypeDispatch-error: " + e.toString()
            )
        }
    }

}